homeassistant/components/fritz/common.py

```diff
@@ -0,0 +1,266 @@
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_DEVICES,
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_USERNAME,
+)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.util import dt as dt_util, get_local_ip
+
+from .const import (
+    ATTR_HOST,
+    DEFAULT_HOST,
+    DEFAULT_PORT,
+    DEFAULT_USERNAME,
+    DOMAIN,
+    ERROR_CONNECTION_ERROR,
+    TRACKER_SCAN_INTERVAL,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+Device = namedtuple("Device", ["mac", "ip", "name"])
+
+
+def ensure_unique_hosts(value):
+    """Validate that all configs have a unique host."""
+    vol.Schema(vol.Unique("duplicate host entries found"))(
+        [socket.gethostbyname(entry[CONF_HOST]) for entry in value]
+    )
+    return value
+
+
+CONFIG_SCHEMA = vol.Schema(
+    vol.All(
+        cv.deprecated(DOMAIN),
+        {
+            DOMAIN: vol.Schema(
+                {
+                    vol.Required(CONF_DEVICES): vol.All(
+                        cv.ensure_list,
+                        [
+                            vol.Schema(
+                                {
+                                    vol.Optional(CONF_HOST): cv.string,
+                                    vol.Optional(CONF_PORT): cv.port,
+                                    vol.Required(CONF_USERNAME): cv.string,
+                                    vol.Required(CONF_PASSWORD): cv.string,
+                                }
+                            )
+                        ],
+                        ensure_unique_hosts,
+                    )
+                }
+            )
+        },
+    ),
+    extra=vol.ALLOW_EXTRA,
+)
+
+SERVICE_SCHEMA = vol.Schema({vol.Required(ATTR_HOST): cv.string})
+
+
+class FritzBoxTools:
+    """FrtizBoxTools class."""
+
+    def __init__(
+        self,
+        hass,
+        password,
+        username=DEFAULT_USERNAME,
+        host=DEFAULT_HOST,
+        port=DEFAULT_PORT,
+    ):
+        """Initialize FritzboxTools class."""
+        # general timeout for all requests to the router. Some calls need quite some time.
+
+        try:
+            self.connection = FritzConnection(
+                address=host, port=port, user=username, password=password, timeout=60.0
+            )
+
+            self.fritzstatus = FritzStatus(fc=self.connection)
+            self._unique_id = self.connection.call_action("DeviceInfo:1", "GetInfo")[
+                "NewSerialNumber"
+            ]
+            self.fritzhosts = FritzHosts(fc=self.connection)
+            self._device_info = self._fetch_device_info()
+            self.success = True
+            self.error = False
+        except FritzConnectionException:
+            self.success = False
+            self.error = ERROR_CONNECTION_ERROR
+
+        self.hass = hass
+        self.ha_ip = get_local_ip()
```

Comment:
We don't allow side effects in `__init__`.  Please move this to `async_setup`

Reply:
Done